### PR TITLE
feat: move non-route related ef configs to function_config in manifest

### PR DESCRIPTION
**Pull request by khendrikse**
…unction_config

🎉 Thanks for sending this pull request! 🎉

Please make sure the title is clear and descriptive.

If you are fixing a typo or documentation, please skip these instructions.

Otherwise please fill in the sections below.

**Which problem is this pull request solving?**

Example: I'm always frustrated when [...]

**List other issues or pull requests related to this problem**

Example: This fixes #5012

**Describe the solution you've chosen**

Example: I've fixed this by [...]

**Describe alternatives you've considered**

Example: Another solution would be [...]

**Checklist**

Please add a `x` inside each checkbox:

- [ ] I have read the [contribution guidelines](../blob/main/CONTRIBUTING.md).
- [ ] The status checks are successful (continuous integration). Those can be seen below.

**A picture of a cute animal (not mandatory but encouraged)**

### CODE REVIEW

File: node/bundler.test.ts [hunk 1]
```diff
@@ -321,7 +321,7 @@ test('Processes a function that imports a custom layer', async () => {
   await cleanup()
 })
 
-test('Loads declarations and import maps from the deploy configuration', async () => {
+test('Loads declarations and import maps from the deploy configuration and in-source-config', async () => {
```
Comment (eduardoboucas) [inline node/bundler.test.ts:324]:
```suggestion
test('Loads declarations and import maps from the deploy configuration and in-source config', async () => {
```

File: node/bundler.ts [hunk 2]
```diff
@@ -173,18 +172,31 @@ const getBasePath = (sourceDirectories: string[], inputBasePath?: string) => {
...
-) => {
-  const fullFuncPath = functions?.find((func) => func.name === declaration.function)?.path
-
+const addMissingConfigFields = ({
+  name,
+  config,
+  functions,
+  declarations,
+  internalFunctionsPath,
+}: {
```
Comment (eduardoboucas) [inline node/bundler.ts:181]: nit: Could we move these options into an interface?

File: node/bundler.ts [hunk 3]
```diff
@@ -173,18 +172,31 @@ const getBasePath = (sourceDirectories: string[], inputBasePath?: string) => {
...
+  name: string
+  config: FunctionConfig
+  functions: EdgeFunction[]
+  declarations: Declaration[]
+  internalFunctionsPath?: string
+}) => {
+  const fullFuncPath = functions?.find((func) => func.name === name)?.path
+  // Check any declarations deprecated name and generator fields
+  const declaration = declarations?.find((decl) => decl.function === name)
```
Comment (eduardoboucas) [inline node/bundler.ts:190]:
```suggestion
  const declaration = declarations?.find((decl) => decl.function === name)

```

File: node/bundler.ts [hunk 4]
```diff
@@ -173,18 +172,31 @@ const getBasePath = (sourceDirectories: string[], inputBasePath?: string) => {
...
 }
 
-export const addGeneratorFieldIfMissing = (
-  declaration: Declaration,
-  functions: EdgeFunction[],
-  internalFunctionsPath?: string,
-) => {
-  const fullFuncPath = functions?.find((func) => func.name === declaration.function)?.path
-
+const addMissingConfigFields = ({
```
Comment (eduardoboucas) [inline node/bundler.ts:175]: I don't think "missing config fields" is very descriptive — what's missing and why? Perhaps we can add a comment explaining why we need this function?

File: node/manifest.ts [hunk 5]
```diff
@@ -43,6 +43,17 @@ interface GenerateManifestOptions {
...
 
+const removeEmptyConfigValues = (functionConfig: EdgeFunctionConfig) =>
+  Object.entries(functionConfig).reduce((acc, [key, value]) => {
+    if (value && !(Array.isArray(value) && value.length === 0)) {
+      return { ...acc, [key]: value }
+    }
+    return acc
+  }, {} as EdgeFunctionConfig)
+
+const hasAnyConfigValues = (functionConfig: EdgeFunctionConfig) =>
```
Comment (eduardoboucas) [inline node/manifest.ts:54]: I don't love the fact that we're defining above what a config object is and then having another allow-list here, so adding a new property will mean adding it in two places. I can see people getting tripped up by that.

I don't have an obvious alternative to offer without getting into more complex schema validation, so I'm happy to defer to you.

File: node/manifest.ts [hunk 6]
```diff
@@ -43,6 +43,17 @@ interface GenerateManifestOptions {
   layers?: Layer[]
 }
 
+const removeEmptyConfigValues = (functionConfig: EdgeFunctionConfig) =>
```
Comment (eduardoboucas) [inline node/manifest.ts:46]: Why do we need to remove empty values?

**Review by eduardoboucas — COMMENTED** [review]
Left a couple of questions before doing a second pass.

File: node/bundler.test.ts (hunk 1 above)
Comment (khendrikse) [reply node/bundler.test.ts:324]: ✅ taking it along in my commit!

File: node/bundler.ts [hunk 7]
```diff
@@ -173,18 +172,31 @@ const getBasePath = (sourceDirectories: string[], inputBasePath?: string) => {
...
+  internalFunctionsPath,
+}: {
+  name: string
+  config: FunctionConfig
+  functions: EdgeFunction[]
+  declarations: Declaration[]
+  internalFunctionsPath?: string
+}) => {
+  const fullFuncPath = functions?.find((func) => func.name === name)?.path
+  // Check any declarations deprecated name and generator fields
```
[the remark this answers is not in the document]
Comment (khendrikse) [reply node/bundler.ts:189]: ✅ taking it along in my commit!

File: node/bundler.ts (hunk 3 above)
Comment (khendrikse) [reply node/bundler.ts:190]: ✅ taking it along in my commit!

File: node/declaration.ts [hunk 8]
```diff
@@ -5,6 +5,7 @@ import { FunctionConfig, Path } from './config.js'
 interface BaseDeclaration {
   cache?: string
   function: string
+  // todo: remove these two after a while and only support in-source-config for non-route related configs
```
[the remark this answers is not in the document]
Comment (khendrikse) [reply node/declaration.ts:8]: ✅ taking it along in my commit!

File: node/bundler.ts (hunk 2 above)
Comment (khendrikse) [reply node/bundler.ts:181]: I shall!

File: node/bundler.ts (hunk 4 above)
Comment (khendrikse) [reply node/bundler.ts:183]: I decided to switch it up a bit and make it, hopefully, more descriptive, and give functions a clear single task.

File: node/manifest.ts (hunk 5 above)
Comment (khendrikse) [reply node/manifest.ts:54]: I tried it this way because [on this line](https://github.com/netlify/edge-bundler/pull/348/files#diff-e659ded745234fd55b239ea6728eef800523583941fd1542d88b44373cd39e8aL56) we were checking some of the values. 

What I could do, is allow for anything in the config to be passed to the function_config. This would mean that we get anything defined in it. Including `cache: manual` for example. Then we only need to filter out any config that is empty.

File: node/manifest.ts (hunk 6 above)
Comment (khendrikse) [reply node/manifest.ts:46]: I did this because the logic before this would allow for undefined values to be passed. For example. if we had `on_error` but `functionConfig.excluded_patterns = []` we would otherwise still get that empty array added, or fields that were undefined. Like `generator: undefined`. Essentially I tried to optimise [this line](https://github.com/netlify/edge-bundler/pull/348/files#diff-e659ded745234fd55b239ea6728eef800523583941fd1542d88b44373cd39e8aL56) so I wouldn't have to add a bunch of other config values. And even if I did, if one of them was true, it would add undefined values for the others.

**Comment by danez** [comment]
So we are not allowing `generator` and `name` being set through the manifest? Why is that?

**Comment by danez** [comment]
FYI: I'm also working on the same code for splitting ISC up into internal vs user. https://github.com/netlify/edge-bundler/pull/347/files

**Comment by khendrikse** [comment]
> So we are not allowing `generator` and `name` being set through the manifest? Why is that?

@danez the idea is that anything that is declared in the declarations for the manifest is mostly meant for routing information. ISC is for configuration on a function level.

**Comment by khendrikse** [comment]
> FYI: I'm also working on the same code for splitting ISC up into internal vs user. https://github.com/netlify/edge-bundler/pull/347/files

Let's first get that merged then!

File: node/bundler.ts [hunk 9]
```diff
@@ -186,19 +183,42 @@ const getBasePath = (sourceDirectories: string[], inputBasePath?: string) => {
...
-  const isInternal = Boolean(internalFunctionsPath && fullFuncPath && isPathInside(fullFuncPath, internalFunctionsPath))
+// Merged any old-style non-in-source configs from declaration.
+const mergeWithDeclarationConfig = ({ name, config, declarations }: MergeWithDeclarationConfigOptions) => {
+  const declaration = declarations?.find((decl) => decl.function === name)
 
-  const generatorFallback = isInternal ? 'internalFunc' : undefined
-  return { ...declaration, generator: declaration.generator || generatorFallback }
+  return {
+    ...config,
+    name: declaration?.name || config.name,
```
Comment (danez) [inline node/bundler.ts:198]: This slightly confused me, until i figured that this is the displayName and the passed in name is the functionName. Could we make this more clear? maybe alias the input `name` as `functionName` or add a comment here that this is the display name

File: node/manifest.ts [hunk 10]
```diff
@@ -54,14 +65,29 @@ const sanitizeEdgeFunctionConfig = (config: Record<string, EdgeFunctionConfig>):
   const newConfig: Record<string, EdgeFunctionConfig> = {}
 
   for (const [name, functionConfig] of Object.entries(config)) {
-    if (functionConfig.excluded_patterns.length !== 0 || functionConfig.on_error) {
-      newConfig[name] = functionConfig
+    const newFunctionConfig = removeEmptyConfigValues(functionConfig)
+
+    if (hasAnyConfigValues(newFunctionConfig)) {
```
Comment (danez) [inline node/manifest.ts:70]: Wouldn't this be the same as `Object.keys(newFunctionConfig).length > 0`?

File: node/manifest.ts [hunk 11]
```diff
@@ -44,6 +44,17 @@ interface GenerateManifestOptions {
   userFunctionConfig?: Record<string, FunctionConfig>
 }
 
+const removeEmptyConfigValues = (functionConfig: EdgeFunctionConfig) =>
+  Object.entries(functionConfig).reduce((acc, [key, value]) => {
+    if (value && !(Array.isArray(value) && value.length === 0)) {
+      return { ...acc, [key]: value }
+    }
+    return acc
+  }, {} as EdgeFunctionConfig)
```
Comment (danez) [inline node/manifest.ts:53]:
```suggestion
  Object.fromEntries(Object.entries(functionConfig).filter(([_, value]) => value && !(Array.isArray(value) && value.length === 0)));
```

this is totally subjective, but I'm not a big fan of reduce, unless it is really reducing something. So I would find this much more readable and probably less objects created, but up to you.

File: node/bundler.ts (hunk 9 above)
Comment (danez) [reply node/bundler.ts:198]: so we could do 
```suggestion
    name,
```
 here?

File: node/manifest.ts (hunk 11 above)
Comment (khendrikse) [reply node/manifest.ts:53]: I think I would find this less readable because it hides a lot of logic and I need to read it a lot of times to understand what is happening. That's why I prefer the reduce option

File: node/bundler.ts (hunk 9 above)
Comment (khendrikse) [reply node/bundler.ts:202]: I changed the way I named it now

File: node/manifest.ts (hunk 10 above)
Comment (khendrikse) [reply node/manifest.ts]: For some reason, before that didn't work... but now it does! Nice catch!

File: node/bundler.ts [hunk 12]
```diff
@@ -186,19 +183,44 @@ const getBasePath = (sourceDirectories: string[], inputBasePath?: string) => {
...
+  ...config,
+  generator: config.generator || 'internalFunc',
+})
+
+interface CreateFunctionConfigOptions {
+  internalFunctionsWithConfig: Record<string, FunctionConfig>
+  declarations: Declaration[]
+}
+
+const createFunctionConfig = ({ internalFunctionsWithConfig, declarations }: CreateFunctionConfigOptions) =>
```
Comment (eduardoboucas) [inline node/bundler.ts:215]: nit: I find this function a bit hard to follow. The name isn't super descriptive — it says it creates function configs but it's iterating one of the parameters and calling two other functions inside. I feel like an abstraction like `addGeneratorFallback`, which is simply adding a field, is adding to the confusion more than the benefits it brings.

File: node/declaration.ts [hunk 13]
```diff
@@ -5,6 +5,7 @@ import { FunctionConfig, Path } from './config.js'
 interface BaseDeclaration {
   cache?: string
   function: string
+  // todo: remove these two after a while and only support in-source config for non-route related configs
```
Comment (eduardoboucas) [inline node/declaration.ts:8]: What's "after a while"? Should this TODO be an issue instead?

File: node/manifest.ts [hunk 14]
```diff
@@ -54,14 +62,29 @@ const sanitizeEdgeFunctionConfig = (config: Record<string, EdgeFunctionConfig>):
...
 
   return newConfig
 }
 
+const addExcludedPatterns = (
+  name: string,
+  manifestFunctionConfig: Record<string, EdgeFunctionConfig>,
+  excludedPath?: Path | Path[],
+) => {
+  if (excludedPath) {
```
Comment (eduardoboucas) [inline node/manifest.ts:80]: nit: You could early-return and avoid nesting all these statements.

```js
if (!excludedPath) {
  return
}
```
